### DISCUSSION TITLE
Alternative zum SDM120 Modbus

### DISCUSSION THREAD

**Discussion by Clarion30** [camueller/SmartApplianceEnabler#369] [Allgemein]
Hallo, 
hätte gerne noch einen Stromzähler leider kann man die im Moment nirgendswo kaufen, finde zumindest nicht so einen.
Hab als Alternative dann einen Onro-we-515  gekauft. mit der vorhanden Software kann ich auch die einzelenen Werte auslesen. 
Aber leider finde ich nicht die einzelnen Register für den Simply Modbus TCP Client, ich kann mich verbinden aber komme bei den Registern dann nicht weiter.
Wie geschrieben würde ich lieber den SDM120 nehmen aber den bekomme ich nicht deswegen die Frage nach einer alternative

Danke für die Hilfe

**Comment by camueller** [comment]
Schau mal hier: https://www.aggsoft.com/serial-data-logger/tutorials/modbus-data-logging/orno-or-we-514-or-we-515.htm

**Comment by Clarion30** [comment]
Ja Hallo danke, da hab ich schon geschaut aber da funktioniert nichts von, zumindest komme ich da nicht weiter :( in der Software ist es das Register hex A000 / Dezimal 40960. der liest dann einmal Total in Kwh aus, sowie T1 bis T4.

**Comment by Clarion30** [comment]
So ich muss nochmal nerven bin jetzt soweit das ich den Totalwert auslesen kann 

![Screenshot 2022-10-20 210739](https://user-images.githubusercontent.com/17002918/197036547-5f8add7d-982a-496e-bd2b-d0b902afefab.jpg)

aber leider nicht im SAE bzw. bleibt es bei 0   müsste den zweiten Wert auslesen (bei 2.0.6 zeigt er auch keine Watt mehr an)

![Screenshot 2022-10-20 211111](https://user-images.githubusercontent.com/17002918/197037290-9f061fbe-1458-4185-8f24-c583bb96fd98.jpg)

gehen doch die SDM einfach ......

Gruß und danke nochmal für die Hilfe
Clarion

**Reply by camueller** [reply]
Der SAE liest aber aus Register 40960 nicht `0, 27` wie im Screenshot, sondern `0, 25`. Probier mal im SAE als Register 960 (also ohne die `40` am Anfang. Du musst es irgendwie schaffen, dass erstmal die gleichen Byte-Werte gelesen werden im SAE wie in Simply Modbus. Danach müssen die Bytes noch zum selben Dezimalwert fürhen, d.h. Du musst im SAE das Format richtig konfigurieren.

**Reply by Clarion30** [reply]
Moin,

Das ist ein Fehler von mir die Screenshots sind zu unterschiedlichen Zeiten gemacht. Die Zahlen sind identisch also dann 0, 27 im SAE sind 270 Watt ist mein Zählerstand

**Reply by camueller** [reply]
Hast Du den Zähler inzwischen im SAE zum Laufen bekommen?

**Reply by Clarion30** [reply]
Nein bin noch im Urlaub erst wieder ab Montag wieder dann sind die Ferien in Hessen rum.Melde mich wenn alles läuft

**Reply by Clarion30** [reply]
so nochmal probiert bekomme es nicht hin bzw. weiß nicht weiter 

![2](https://user-images.githubusercontent.com/17002918/198965640-7f3eaa91-cea5-4515-999e-f77280debe1e.jpg)

152 sind bei mir 1,52 Kwh das passt im Sunnyportal zeigt er mir nur 0 an ?

**Reply by camueller** [reply]
Wenn 1,52 kWh Dein Zählerstand laut Zählerdisplay ist und Du beim Auslesen im SAE 152 bekommst, musst Du lediglich als Umrechnungsfaktor 0.001 im SAE einstellen. Falls der Zähler keine genaueren Werte mit mindestens 5 Nachkommstellen liefert, wirst Du vermutlich damit nicht glücklich werden. Oder Du liesst nicht den Zählerstand aus, sondern direkt die Leistung (also anderes Register).

**Reply by Clarion30** [reply]
nö geht aucht nicht als Leistung Register 320 (60W Glühbirne) sowie Register 40960 als Zählerstand bleibt bei null :( im Sunny Portal.
In der shell zeigt er die Werte richtig an Fehler liegt hier anscheinend zwischen den Ohren  :)

**Reply by camueller** [reply]
Wenn Du willst, kannst Du mir den Zähler per Post schicken, dann teste ich mal, wie man den mit dem SAE zum Laufen bekommt. Allerdings bin ich ab Wochende für 14 Tage weg, kann mich also erst danach darum kümmern.

**Reply by Clarion30** [reply]
Kein Problem brauche deine Adresse dann schick ich ihn dir zu, in 2 Wochen oder sofort wie Du es gerne hättest :).
Danke Dir erstmal für deine Bemühungen.

**Comment by Clarion30** [comment]
Würde heute rausschicken wenn in Ordnung ? Lege meine Adresse sowie email bei, Geld für Rücksendung und eine kleine Spende für das Project wenn es dir Recht ist ?.

Gruß Clarion30

**Reply by camueller** [reply]
Ja, wobei Du das Geld vielleicht lieber überweisen solltest (Paypal/Überweisung siehe Homepage) - dann kommt es nicht weg :-)

**Comment by Clarion30** [comment]
Okay Geld ist raus per paypal sowie Paket ist unterwegs

**Reply by camueller** [reply]
Das Geld ist schon mal da - vielen Dank! Ich halte Dich auf dem Laufenden, was die Tests mit dem Zähler angehen.

**Reply by camueller** [reply]
Zähler ist heute angekommen.

**Comment by Clarion30** [comment]
Super

25.11.2022 16:01:38 Axel Müller ***@***.***>:

**Comment by camueller** [comment]
Habe mir mal die [Modbus-Registerbeschreibung](https://b2b.orno.pl/download-resource/26063/) heruntergeladen.
Der Zählerstand wird als Ganzzahlwert (LONG) geliefert in Hundertstel-kWh.

![image](https://user-images.githubusercontent.com/4489628/204014537-d1b8980c-b9af-4d7c-89bf-611a64663c49.png)

Der SAE benötigt 5-Nachkommastellen, um die Leistung korrekt zu berechnen, aber dieser Zähler liefert nur 2 Nachkommastellen. Damit bleibt bei diesem Zähler nur die Möglichkeit, die Leistung direkt abzufragen (anstatt sie aus Zählerstandsdifferenzen zu berechnen).

**Comment by Clarion30** [comment]
Das wäre ja auch nicht schlimm, mir wurde Zählerstand sowie Leistung nicht angezeigt

25.11.2022 16:23:52 Axel Müller ***@***.***>:

**Comment by camueller** [comment]
Wenn ich das Auslesen von Registern eines Modbus-Gerätes teste, mache ich das gern mit dem einfachen Kommandline-Tool `mbpoll`, das auch in der [Dokumentation](https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/Modbus_DE.md) beschrieben ist.

Damit läßt sich das grundsätzliche Funktionieren des Protokolls und auch  verschiedene Einstellungen wesentlich schneller testen, als wenn man das im SAE macht.

# Zählerstand (in 0,01 kWh also hier 2,5 kWh) 
```
pi@raspberrypi:~ $ mbpoll -a 1 -r 40960 -0 -1 -t 4:int -B localhost
mbpoll 1.0-0 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright © 2015-2019 Pascal JEAN, https://github.com/epsilonrt/mbpoll
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions; type 'mbpoll -w' for details.

Protocol configuration: Modbus TCP
Slave configuration...: address = [1]
                        start reference = 40960, count = 1
Communication.........: localhost, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 32-bit integer (big endian), output (holding) register table

-- Polling slave 1...
[40960]:        250
```

# Leistung (in W, also hier 30W mit meinem angeschlossenen 30W-Lötkolben) 
```
pi@raspberrypi:~ $ mbpoll -a 1 -r 320 -0 -1 -t 4:int -B localhost
mbpoll 1.0-0 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright © 2015-2019 Pascal JEAN, https://github.com/epsilonrt/mbpoll
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions; type 'mbpoll -w' for details.

Protocol configuration: Modbus TCP
Slave configuration...: address = [1]
                        start reference = 320, count = 1
Communication.........: localhost, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 32-bit integer (big endian), output (holding) register table

-- Polling slave 1...
[320]:  30
```

**Comment by camueller** [comment]
Wenn man dann weiss, welche Werte funktionieren, kann man diese im SAE einstellen:

![image](https://user-images.githubusercontent.com/4489628/204074055-68d462de-6441-4bb6-aade-f0880fe4a6bc.png)

und findet danach im Log die Einträge, die zeigen, dass auch der SAE die Leistung korrekt auslesen kann:

```
2022-11-26 06:38:00,780 DEBUG [Timer-0] d.a.s.m.ModbusSlave [ModbusSlave.java:77] F-00000001-000000000002-00: Connecting to modbus modbus@127.0.0.1:502
2022-11-26 06:38:00,782 TRACE [Timer-0] d.a.s.m.e.ReadHoldingRegisterExecutor [ReadHoldingRegisterExecutor.java:38] F-00000001-000000000002-00: Reading holding register=320 requestWords=2
2022-11-26 06:38:00,852 DEBUG [Timer-0] d.a.s.m.e.ReadHoldingRegisterExecutor [ReadHoldingRegisterExecutor.java:52] F-00000001-000000000002-00: Holding register=320 value=[0, 31]
2022-11-26 06:38:00,853 DEBUG [Timer-0] d.a.s.m.t.FloatValueTransformer [IntegerValueTransformer.java:43] F-00000001-000000000002-00: transformed value=31
```

**Comment by camueller** [comment]
@Clarion30 Ich schicke Dir den Zähler heute wieder zurück.

**Comment by Clarion30** [comment]
Moin wieder da, teste es nochmal bei mir danke erstmal.
